omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+            return -1;
+          }
+          thisOffset = structSize+lastOffset;
+        }
+        catch (...)
+        {
+          structInvalid = true;
+          std::cout<<"Failed to find the standard datatype: " << dtype << ". Definition for " << structName << " and its dependents failed" << std::endl;
+          return -1;
+        }
+      }
+    }
+    if (offSetsCounted < row.size())
+    {
+      newRow.push_back(thisOffset);
+    }
+    else
+    {
+      // if this is the last offset, then do not add it to the row as it points to the end of the struct
+      // instead we return thisOffset
+      if (!structInvalid)
+      {
+        structMap[structName] = newRow;
+        return thisOffset;
+      }
+      else
+        return -1;
+    }
+  }
+  // Dont think its possible to get here
+  return -1;
+}
+
+std::string drvOmronEIP::findNextDtype(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::string structName)
+{
+  try
+  {
+    std::vector<std::string> row = rawMap.at(structName);
+    for (std::string dtype : row)
+    {
+      if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL" || dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD" || dtype =="ULINT" || dtype =="LINT" || dtype == "LREAL" || dtype =="LWORD" || dtype.substr(0,6) == "STRING") {
+        return dtype;
+      }
+      else {
+        // We need to find the next dtype within an embedded structure
+        std::string nextDtype = findNextDtype(rawMap, dtype);
+        return nextDtype;
+      }
+    }
+  }
+  catch (...)
+  {
+    if (structName =="UINT" || structName =="INT" || structName =="WORD" || structName =="BOOL" || structName =="UDINT" || structName =="DINT" || structName =="REAL" || structName =="DWORD" || structName =="ULINT" || structName =="LINT" || structName == "LREAL" || structName =="LWORD" || structName.substr(0,6) == "STRING" ){
+      return structName;
+    }
+    else {
+      return "structName is neither a structure or a standard dtype";
+    }
+  }
+  return "no dtypes left";
+}
+
+void drvOmronEIP::readPoller()
+{
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int offset;
+  int status;
+  int still_pending = 1;
+  int timeTaken = 0;
+  //The poller may not be fully initialised until the startPollers_ flag is set to 1, do not attempt to get the pPoller yet
+  while (!this->startPollers_ && !omronExiting_) {epicsThreadSleep(0.1);}
+  omronEIPPoller* pPoller = pollerList_.at(threadName);
+  double interval = pPoller->updateRate_;
+  asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Starting poller: %s with interval: %f\n", driverName, functionName, threadName.c_str(), interval);
+  while (!omronExiting_)
+  {
+    double waitTime = interval - ((double)timeTaken / 1000);
+    if (waitTime >=0) {
+      epicsThreadSleep(waitTime);
+    }
+    else {
```

Comment:
This bit here is another candidate for using epicsEvent. I would create an epicsEvent called `omronExiting_` and change the while loop to `while(true)`. Then instead of the `epicsThreadSleep` I would use a the `omronExiting_.wait(waitTime)` (probably works also with waitTime = 0) and IF the event has been signalled, just `break` out of the while loop.

Reply:
https://epics.anl.gov/base/R3-14/12-docs/AppDevGuide.pdf p.289

Reply:
This has the same problem as the startPollers_ event, each poller needs to see this event so that it can shut down.